### ISSUE TITLE
eml_get --replacement & EML validate

### ISSUE THREAD

**Issue by lwasser** [ropensci/EML#139]
Hi again @cboettig 
eml_validate is hanging on the file:
```
eml_HARV <- read_eml("http://harvardforest.fas.harvard.edu/data/eml/hf001.xml")
eml_validate(eml_HARV)
```
Any thoughts on this?

Also previously i was using eml_get as follows to extract key metadata components. What is the suggested workflow now given the most recent updates?

```
# get contact
eml_get(eml_HARV,"contact")

# grab all keywords in the file
eml_get(eml_HARV,"keywords")

# figure out the extent & temporal coverage of the data
eml_get(eml_HARV,"coverage")
```

Would i use a slot approach with `@` signs? Thank you.
eml_HARV@dataset - is hanging significantly as well.

**Comment by cboettig** [comment]
Re: `eml_validate()`, works fine for me, not sure why it should hang.  `eml_validate()` is currently just doing schema validation, so it's just a thin wrapper around `XML::xmlSchemaValidate()`.  May be related to #138; do try a clean restart.  

`eml_HARV@dataset` takes a long time to print for me too (~2 minutes!), though doesn't hang.  We'll have to look into speeding that up.  Note that the actual subsetting is fast, it is just the "show" method that shows the resulting eml information using the XML layout that is so slow.  i.e. subsetting a smaller chunk of EML will actually work much faster (< 3 seconds):  

`eml_HARV@dataset@dataTable[[1]]`


Re `eml_get`, the difficulty with the original function is that many of these elements can appear in lots of different places. Thus `eml_get(eml, "coverage")` is potentially ambiguous; since coverage elements can appear in a lot of places in EML.  

My current thought is to start by writing the low-level functions in the package, and then add higher-level functions that build on these.  Like you say, using `@` directly is low-level, but very general.  For simple object types, it will be about as concise as any helper function anyway, e.g. 

```r
eml_HARV@dataset@contact
```

specifies exactly which contact node we want.  For certain metadata, you often want it returned all together, and probably not as a list or S4 object.  So far, I've implemented this for the attributeList, e.g. for your example, try:

```
get_attributes(eml_HARV@dataset@dataTable[[3]]@attributeList, join = TRUE)
```

Note we use a combination of `@` and the helper function `get_attributes()` here.  The former helps specify exactly which `dataTable` we are talking about (the third one in the dataset).  The `get_attributes()` function turns the big attributesList S4 object into a data.frame, showing much of the metadata available there (though this example leaves a lot of metadata fields as missing).  

So far, there is no `get_coverage()` function or other `get` methods.  I first have to figure out what is the "right" format that such a method should return.  In general I think these helper methods should return R data.frames, but mapping from the XML to a data.frame can be somewhat subjective, since the XML objects are often very highly nested.  Very open to suggestions for this, I'll be opening a few issues outlining some ideas for common elements like `coverage`.

**Comment by lwasser** [comment]
thank you so much, @cboettig for addressing these questions! i'm trying to understand how slots have changed so I can better access elements of the file.

In the file above, how do I access x and y coordinates? it seems as if "indexing" by slot has changed too. When i try to find "coverage" as a slot, the options that i have after

`eml_HARV@dataset`

Have changed. How do i access coverage information now using slots to say get a specific lat/long value. DO i use `get_attributes()`? And if so, how do i find the slot / index level that i need without going back to the file itself.

```
# grab x coordinate - the old code that no longer works
XCoord <- eml_HARV@dataset@coverage@geographicCoverage@boundingCoordinates@westBoundingCoordinate
# grab y coordinate
YCoord <- eml_HARV@dataset@coverage@geographicCoverage@boundingCoordinates@northBoundingCoordinate
```

thank you, Carl!

**Comment by cboettig** [comment]
Short answer: You now need one more slot in your chain: `eml_HARV@dataset@ResourceGroup@coverage@...`

Admittedly, this was perhaps an unwise change, because just looking at the XML you don't see anything called "ResourceGroup".  A "ResourceGroup" is a collection of slots that used to all be direct slots in a `dataset`, but are now all grouped together as child slots.  You can get a sense of how this works by looking at [this diagram](https://github.com/ropensci/EML/blob/master/inst/xsd/docs/eml-2.1.1/eml-dataset.png):

![](https://raw.githubusercontent.com/ropensci/EML/master/inst/xsd/docs/eml-2.1.1/eml-dataset.png)


This shows that 

- (A) an EML `dataset` object is super rich and thus super complicated, and 
- (B) that a `ResourceGroup` is a collection of common slots, including `coverage`.  A `dataset` is not the only place in EML where `ResourceGroup`s appear, so it helps to have a common name for that group of slots.

By grouping the slots in this way, it becomes a bit easier to re-use or mix & match groups of elements when creating your own EML.  (Technical comment: Previously, the slots of a ResourceGroup were inherited using `contains` in the S4 class, rather than creating a slot called ResourceGroup.  That's a bit more natural choice technically speaking, and it still allows a ResourceGroup to be turned into a `dataset` using coercion (as), but it doesn't permit the use case of adding an existing ResourceGroup set of slots to a `dataset` that already has other data.  Still, perhaps we should return to that original layout to avoid this confusion.  Maybe @mbjones can weigh in on that).  

Note that you can see all possible children of, say, a `dataset` object using `getSlots("dataset")`.

**Comment by mbjones** [comment]
We should discuss whether ResourceGroup (and other groups) should really be in the object model.  It does not show up in the XML documents, and is really only an XSD schema construct so that a reusable set of elements can be defined.  I think I would argue that we shouldn't have the ResourceGroup in the R class at all, but instead have slots for all of the elements contained in the ResourceGroup.  Then the R structure would be parallel to the EML structure, and make it consistent.  Thoughts @cboettig and @lwasser ?

**Comment by cboettig** [comment]
@mbjones @lwasser yup, very good point.  For clarity, this is now it's own issue: #141

**Comment by lwasser** [comment]
@mbjones @cboettig i'll have a look at the new issue. from a user perspective it's very confusing to use a slot that i can't see in the actual xml file. it also is confusing as an educator to explain to people. i'd lean towards not using it but will read the other issue more carefully first!

**Comment by cboettig** [comment]
Okay, I think this is now resolved now that #141 is closed?  Feel free to re-open or open a new issue if I've missed things (more than likely)...
